### ISSUE TITLE
AUC& Loss not converage on MNIST

### ISSUE THREAD

**Issue by chenyuqi990215** [SeldonIO/alibi-detect#209]
I use the following code to train model on MNIST using pytorch, and use onnx-tf to convert the model from pytorch to tensorflow in order to use your code. I follow your paper to construct the auto-encoder. However, during training the loss does not coverage (stuck at about 17.36) and I use the universal perturbation generated to test your model, however, the AUC I get is quite low, could you please help me find the bug.

Following is the code to train on MNIST:

```python
# -*-coding:utf-8-*-
import torch
import torch.nn as nn
import torchvision
import torchvision.transforms as transforms

device = torch.device('cpu' if torch.cuda.is_available() else 'cpu')

num_epochs = 100
num_classes = 10
batch_size = 100
learning_rate = 0.001


class LeNet5(nn.Module):
    def __init__(self, num_clases=10):
        super(LeNet5, self).__init__()

        self.c1 = nn.Sequential(
            nn.Conv2d(1, 6, kernel_size=5, stride=1, padding=2),
            # nn.BatchNorm2d(6),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2)
        )

        self.c2 = nn.Sequential(
            nn.Conv2d(6, 16, kernel_size=5),
            # nn.BatchNorm2d(16),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2)
        )

        self.c3 = nn.Sequential(
            nn.Conv2d(16, 120, kernel_size=5),
            # nn.BatchNorm2d(120),
            nn.ReLU()
        )

        self.fc1 = nn.Sequential(
            nn.Linear(120, 84),
            nn.ReLU()
        )

        self.fc2 = nn.Sequential(
            nn.Linear(84, num_classes),
            # nn.LogSoftmax()
        )

        self.features = None

    def forward(self, x):
        out = self.c1(x)
        out = self.c2(out)
        out = self.c3(out)
        self.features = out
        out = out.reshape(out.size(0), -1)
        out = self.fc1(out)
        out = self.fc2(out)
        return out

def train():
    # MNIST dataset
    train_dataset = torchvision.datasets.MNIST(root='data/',
                                               train=True,
                                               transform=transforms.ToTensor(),
                                               download=True)

    test_dataset = torchvision.datasets.MNIST(root='data/',
                                              train=False,
                                              transform=transforms.ToTensor())

    # Data loader
    train_loader = torch.utils.data.DataLoader(dataset=train_dataset,
                                               batch_size=batch_size,
                                               shuffle=True)

    test_loader = torch.utils.data.DataLoader(dataset=test_dataset,
                                              batch_size=batch_size,
                                              shuffle=False)

    model = LeNet5(num_classes).to(device)

    criterion = nn.CrossEntropyLoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)

    best_model = LeNet5()
    best_acc = 0

    total_step = len(train_loader)
    for epoch in range(num_epochs):
        model.train(True)
        for i, (images, labels) in enumerate(train_loader):
            images = images.to(device)
            labels = labels.to(device)

            # Forward pass
            outputs = model(images)
            loss = criterion(outputs, labels)

            # Backward and optimize
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            if (i + 1) % 100 == 0:
                print('Epoch [{}/{}], Step [{}/{}], Loss: {:.4f}'
                      .format(epoch + 1, num_epochs, i + 1, total_step, loss.item()))

        model.train(False)

        with torch.no_grad():
            correct = 0
            total = 0
            for images, labels in test_loader:
                images = images.to(device)
                labels = labels.to(device)
                outputs = model(images)
                _, predicted = torch.max(outputs.data, 1)
                total += labels.size(0)
                correct += (predicted == labels).sum().item()

            print('Epoch {} Accuracy of the network: {}%'.format(epoch,100 * correct / total))  # 输出识别准确率
            if correct > best_acc:
                best_acc = correct
                best_model.load_state_dict(model.state_dict())

    torch.save(best_model.state_dict(), 'model/mnist_lenet5_v1.pkl')

if __name__ == "__main__":
    train()
```

Following is the code for training and Inferencing your model

```python
class Alibi(Detector):
    def __init__(self):
        super(Alibi, self).__init__("Alibi")
        if DATASET_NAME == "CIFAR-10":
            latent_dim = 40
            self.encoder_net = tf.keras.Sequential(
                [
                    tf.keras.layers.InputLayer(input_shape=(32, 32, 3)),
                    tf.keras.layers.Conv2D(64, 4, strides=2, padding='same', activation=tf.nn.relu),
                    tf.keras.layers.Conv2D(128, 4, strides=2, padding='same', activation=tf.nn.relu),
                    tf.keras.layers.Conv2D(512, 4, strides=2, padding='same', activation=tf.nn.relu),
                    tf.keras.layers.Flatten(),
                    tf.keras.layers.Dense(latent_dim)
                ])
            self.decoder_net = tf.keras.Sequential(
                [
                    tf.keras.layers.InputLayer(input_shape=(latent_dim,)),
                    tf.keras.layers.Dense(4 * 4 * 128),
                    tf.keras.layers.Reshape(target_shape=(4, 4, 128)),
                    tf.keras.layers.Conv2DTranspose(256, 4, strides=2, padding='same', activation=tf.nn.relu),
                    tf.keras.layers.Conv2DTranspose(64, 4, strides=2, padding='same', activation=tf.nn.relu),
                    tf.keras.layers.Conv2DTranspose(3, 4, strides=2, padding='same', activation='sigmoid')
                ])
            (x, y), (x_test, y_test) = keras.datasets.cifar10.load_data()
            x = (x.astype(np.float32) / 255)
            x_test = (x_test.astype(np.float32) / 255)
            print(x.shape)
        elif DATASET_NAME == "MNIST":
            latent_dim = 10
            self.encoder_net = tf.keras.Sequential(
                [
                    tf.keras.layers.InputLayer(input_shape=(28, 28, 1)),
                    tf.keras.layers.Conv2D(64, 4, strides=2, padding='same', activation=tf.nn.relu),
                    tf.keras.layers.Conv2D(128, 4, strides=2, padding='same', activation=tf.nn.relu),
                    tf.keras.layers.Conv2D(512, 4, strides=2, padding='same', activation=tf.nn.relu),
                    tf.keras.layers.Flatten(),
                    tf.keras.layers.Dense(latent_dim)
                ])
            self.decoder_net = tf.keras.Sequential(
                [
                    tf.keras.layers.InputLayer(input_shape=(latent_dim,)),
                    tf.keras.layers.Dense(7 * 7 * 32),
                    tf.keras.layers.Reshape(target_shape=(7, 7, 32)),
                    tf.keras.layers.Conv2DTranspose(64, 4, strides=2, padding='same', activation=tf.nn.relu),
                    tf.keras.layers.Conv2DTranspose(32, 4, strides=2, padding='same', activation=tf.nn.relu),
                    tf.keras.layers.Conv2DTranspose(1, 4, strides=1, padding='same', activation='sigmoid')
                ])
            (x,y), (x_test, y_test) = keras.datasets.mnist.load_data()
            x = x.reshape(-1, 28, 28, 1).astype(np.float32) / 255
            x_test = x_test.reshape(-1, 28, 28, 1).astype(np.float32) / 255
        else:
            raise NotImplementedError
        self.ad = AdversarialAE(
                encoder_net=self.encoder_net,
                decoder_net=self.decoder_net,
                model=self.pytorch_to_keras(build_model(DATASET_NAME,MODEL_NAME)),
            )
        self.ad.fit(x,epochs=10, batch_size=256,verbose=True)

        # np.save("./data/result/normal.txt",self.ad.score(x_test,batch_size=64))
        self.ad.infer_threshold(x_test, threshold_perc=85, margin=0., batch_size=64)
        self.x_test = x_test


    def pytorch_to_keras(
            self, model, input_shapes=None,
            change_ordering=True, verbose=False, name_policy=None,
    ):
        input = torch.randn([100, *image_size(DATASET_NAME)])
        # 设置输入张量名，多个输入就是多个名
        input_names = ["input"]
        # 设置输出张量名
        output_names = ["output"]
        # 自定义onnx文件名和路径
        onnx_filename = "model.onnx"
        # 执行转化和保存
        torch.onnx.export(model, input, onnx_filename, verbose=True, input_names=input_names,
                          output_names=output_names)

        onnx_model = onnx.load(onnx_filename)

        k_model = onnx_to_keras(onnx_model=onnx_model, input_names=["input"],
                                input_shapes=input_shapes, name_policy=name_policy,
                                verbose=verbose, change_ordering=change_ordering)
        k_model.compile()
        return k_model

    def query(self,images, threshold=None):
        images = images.cpu().numpy()
        images = images.transpose(0,2,3,1)
        x_adv = images
        return np.sum(self.ad.score(x_adv, batch_size=100) <= self.ad.threshold)

    def save(self, pert, file_path):
        pert = pert.reshape((1,*image_size(DATASET_NAME)))
        x_adv = (self.x_test + pert.transpose(0,2,3,1)).astype(np.float32)
        np.save(file_path, self.ad.score(x_adv, batch_size=64))
```

The Universal Perturbation generated by my code can be found in the following link:
https://pastebin.ubuntu.com/p/jVQjRZGPv6/

Thank you very much!
